### DISCUSSION TITLE
Background Sync

### DISCUSSION THREAD

**Discussion by s-cork** [anvilistas/anvil-labs#85] [Ideas]
https://github.com/anvilistas/anvil-labs/discussions/81#discussioncomment-3408873 - @Coastline23

> Hope it's ok to ask this here as it's pretty closely related to the discussion and work already done.
> 
> @s-cork how difficult would it be to use a [PWA Service worker background sync](https://developer.mozilla.org/en-US/docs/Web/API/Background_Synchronization_API) for this sort of update?
> 
> That way the update should be able to cope with a loss of internet and still update even if the app is closed/no longer in foreground or the device is off(asleep) once internet returns.
> 
> The App I'm developing requires a messaging function so the ability to enter a message when offline click send put the phone back in your pocket and have the message sent once the phone gets a connection is much preferable to only sending once you go back to the app (possibly hours or days later).
> 
> This would truly make the offline usability of Anvil awesome!!

**Comment by s-cork** [comment]
Moved the discussion here since it's a separate topic

> how difficult would it be to use a [PWA Service worker background sync](https://developer.mozilla.org/en-US/docs/Web/API/Background_Synchronization_API)

The web worker stuff took me a while and was just to explore what was possible.
I wanted to be able to write a module in python that could be used as the code inside a webworker context.
To create a python wrapper around the service worker api would take just as much, if not more, time.

It's possible to do this now with javascript but it'll take some time to work out how to hack it together.
You also won't have access to anvil things, like `anvil.server.call` so you'd need to use the javascript `fetch` api with anvil http endpoints on the server.
As well as communicating between service-worker/browser via the messages interface.

It's worth noting that this is only really possible in chrome/android. 
The background sync is not available in firefox/safari/ios.

**Sponsorship**
@meatballs and I are open to sponsorship as a means to build certain tools.
Otherwise we build things for ourselves, or other clients.
Time is the limiting factor generally!

Happy to give hints and guide you to ways to build it, but it's not something I'll get to imminently.
The first step would be working out how to register a service worker.
Then taking inspiration from this javascript example for your messaging app: https://github.com/jakearchibald/emojoy

**Reply by Coastline23** [reply]
Hi @s-cork,

Thanks for the response.

I completely understand the lack of time issue.

Your comments have inspired me to take a deeper look at it myself.  

I was initially concerned that the background service worker registration had to be done at the same time as the main PWA Service worker install.  Which I didn't think we could change and I wouldn't want to mess with that.  I think that put me off looking to deeply into it.

However it looks like I can just register additional services to the main worker.  I guess I just put the Javascript for that in Native Libraries like I have done with other things.

I'm going to have a look at it and let you know how I get on.  If I get super stuck with something I'll take you up on the kind offer of guidance and hints if that's ok.

I'd already discussed the possibility of a paid option with @meatballs through the Anvil Forum so I'm not averse to that option either.  Could you give me some idea of the Sponsorship cost required to develop a Javascript only demo implementation of background sync used from within Anvil.

I have already setup the push notification side using OneSignal.

I'm actually OK with a messaging system that just informs the user that the message cannot be sent due to no connection and that it will remain in draft until they try again.  It's clunky I know but in our use case it would be adequate.  But obviously if I can get some form of background sync working then it becomes harder to tell a PWA app from a native one which is obviously the goal.

I'm also ok with support for background sync being limited to Chrome/Edge as I can specify a minimum use requirement to the app as it's for internal company team use.

And finally, just so you know, the work that has been done in Anvil Extras, the Anvil Labs and the help that you all provide to others like me in the Anvil forum is awesome.  You should all feel very proud of it.

Thanks

Tony

**Reply by s-cork** [reply]
Sponsorship is just asking one of me or @meatballs to commit our time to implementing a particular feature. We'd give you a time estimate and a rate and then go from there. 

Great. Yeah you should be able to register the service worker in native libraries. You can probably put the code for the service worker in a js file in your assets. You may need to play with the script url in the `.register()` method to get it to work across published and debug environments.
